File: combinatorics/resource/ST/project_5/bai_toan_6/backtracking.cpp
Tree Edit Distance bang Backtracking (C++)}]
#include <iostream>               // Thu vien nhap xuat
#include <vector>                 // Dung vector de luu cac con cua cay
#include <string>                 // Xu ly chuoi
#include <algorithm>              // Dung ham min
using namespace std;

// Dinh nghia mot nut cay
struct TreeNode {
    string label;                 // Nhan cua nut
    vector<TreeNode*> children;  // Danh sach con
    TreeNode(string l) : label(l) {}  // Ham tao khoi tao nhan
};

// Ham tinh khoang cach giua hai cay
int treeEditDistance(TreeNode* t1, TreeNode* t2) {
    if (!t1 && !t2) return 0;     // Ca hai cay rong -> khoang cach = 0
    if (!t1) {                    // Neu t1 rong -> chen tat ca nut cua t2
        int cost = 1;            // Chen nut goc
        for (auto c : t2->children)
            cost += treeEditDistance(nullptr, c); // De quy chen con
        return cost;
    }
    if (!t2) {                    // Neu t2 rong -> xoa tat ca nut cua t1
        int cost = 1;            // Xoa nut goc
        for (auto c : t1->children)
            cost += treeEditDistance(c, nullptr); // De quy xoa con
        return cost;
    }

    // Chi phi doi nhan (neu khac nhau)
    int cost = (t1->label == t2->label) ? 0 : 1;

    int m = t1->children.size(), n = t2->children.size(); // So con
    int min_cost = INT32_MAX;     // Dat chi phi ban dau vo cung

    // De quy ghep cac nut con
    function<void(int,int,int)> dfs = [&](int i, int j, int acc) {
        if (i == m && j == n) {   // Neu da ghep het ca hai danh sach con
            min_cost = min(min_cost, acc); // Cap nhat ket qua nho nhat
            return;
        }
        if (i < m && j < n)
            // Ghep con thu i cua t1 voi con thu j cua t2
            dfs(i+1, j+1, acc + treeEditDistance(t1->children[i], t2->children[j]));
        if (i < m)
            // Xoa con thu i cua t1
            dfs(i+1, j, acc + treeEditDistance(t1->children[i], nullptr));
        if (j < n)
            // Chen con thu j cua t2
            dfs(i, j+1, acc + treeEditDistance(nullptr, t2->children[j]));
    };
    dfs(0, 0, 0); // Bat dau ghep tu dau danh sach con

    return cost + min_cost; // Tong chi phi = chi phi doi nhan + ghep con
}

// Ham nhap mot cay tu ban phim
TreeNode* inputTree() {
    string label;
    cout << "Nhap nhan goc: ";
    cin >> label;
    TreeNode* node = new TreeNode(label);   // Tao nut moi
    int n;
    cout << "So con cua " << label << ": ";
    cin >> n;
    for (int i = 0; i < n; ++i) {
        cout << "-- Con thu " << i+1 << " cua " << label << ":\n";
        node->children.push_back(inputTree()); // Nhap de quy con
    }
    return node;
}

// Ham main
int main() {
    cout << "Nhap cay 1:\n";
    TreeNode* t1 = inputTree();  // Nhap cay thu nhat
    cout << "Nhap cay 2:\n";
    TreeNode* t2 = inputTree();  // Nhap cay thu hai
    cout << "Edit distance: " << treeEditDistance(t1, t2) << endl; // In ket qua
    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_6/backtracking.py
# Lop cay voi nhan va danh sach con
class TreeNode:
    def __init__(self, label):
        self.label = label                 # Luu nhan cua nut
        self.children = []                # Khoi tao danh sach con rong

# Ham tinh khoang cach chinh sua giua hai cay bang backtracking
def tree_edit_distance(t1, t2):
    if t1 is None and t2 is None:         # Neu ca hai cay rong
        return 0                          # -> khoang cach la 0
    if t1 is None:                        # Neu chi co t1 rong
        # Phai chen nut t2 va tat ca con cua no
        return 1 + sum(tree_edit_distance(None, child) for child in t2.children)
    if t2 is None:                        # Neu chi co t2 rong
        # Phai xoa nut t1 va tat ca con cua no
        return 1 + sum(tree_edit_distance(child, None) for child in t1.children)
    
    # Neu nhan giong nhau thi chi phi la 0, nguoc lai la 1
    cost = 0 if t1.label == t2.label else 1

    # Lay so luong con cua hai nut
    m, n = len(t1.children), len(t2.children)
    min_cost = float('inf')              # Dat chi phi nho nhat ban dau vo cung

    # Ham de quy thu tat ca cach ghep con giua hai cay
    def dfs(i, j, acc):                  # acc: tong chi phi hien tai
        nonlocal min_cost
        if i == m and j == n:            # Neu da ghep het tat ca con
            min_cost = min(min_cost, acc) # Cap nhat chi phi nho nhat
            return
        if i < m and j < n:
            # Ghep con thu i cua t1 voi con thu j cua t2
            dfs(i+1, j+1, acc + tree_edit_distance(t1.children[i], t2.children[j]))
        if i < m:
            # Xoa con thu i cua t1
            dfs(i+1, j, acc + tree_edit_distance(t1.children[i], None))
        if j < n:
            # Chen con thu j cua t2
            dfs(i, j+1, acc + tree_edit_distance(None, t2.children[j]))
    
    dfs(0, 0, 0)                         # Bat dau ghep tu con dau tien cua hai nut
    return cost + min_cost               # Tong chi phi = chi phi doi nhan + ghep con

# Ham nhap cay tu ban phim
def input_tree():
    label = input("Nhap nhan goc: ")     # Nhap nhan cua nut hien tai
    node = TreeNode(label)               # Tao nut moi
    n = int(input(f"So con cua {label}: ")) # Nhap so luong con
    for i in range(n):
        print(f"-- Con thu {i+1} cua {label}:")
        child = input_tree()             # Nhap de quy nut con
        node.children.append(child)      # Them con vao danh sach
    return node                          # Tra ve cay vua nhap

# Chuong trinh chinh
if __name__ == "__main__":
    print("Nhap cay 1:")
    t1 = input_tree()                    # Nhap cay thu nhat
    print("Nhap cay 2:")
    t2 = input_tree()                    # Nhap cay thu hai
    print("Edit distance:", tree_edit_distance(t1, t2))  # Tinh va in ket qua